render_delegate/instancer.cpp

```diff
@@ -76,106 +87,173 @@ void HdArnoldInstancer::_SyncPrimvars()
         }
     }
 
-    // TODO(pal): support motion blur.
-
-    if (GetParentId().IsEmpty()) {
-        return transforms;
+    const auto parentId = GetParentId();
+    if (parentId.IsEmpty()) {
+        return;
     }
 
-    auto* parentInstancer =
-        dynamic_cast<HdArnoldInstancer*>(GetDelegate()->GetRenderIndex().GetInstancer(GetParentId()));
-    if (!TF_VERIFY(parentInstancer)) {
-        return transforms;
+    auto* parentInstancer = dynamic_cast<HdArnoldInstancer*>(GetDelegate()->GetRenderIndex().GetInstancer(parentId));
+    if (ARCH_UNLIKELY(parentInstancer == nullptr)) {
+        return;
     }
 
-    const auto parentTransforms = parentInstancer->CalculateInstanceMatrices(id);
-
-    const auto numParentInstances = parentTransforms.size();
-    if (numParentInstances == 0) {
-        return transforms;
+    HdArnoldSampledMatrixArrayType parentMatrices;
+    parentInstancer->CalculateInstanceMatrices(id, parentMatrices);
+    if (parentMatrices.count == 0 || parentMatrices.values.front().empty()) {
+        return;
     }
 
-    if (numParentInstances > 1) {
-        transforms.resize(numInstances * numParentInstances);
-    }
+    HdArnoldSampledMatrixArrayType childMatrices{sampleArray};
+    _AccumulateSampleTimes(parentMatrices, sampleArray);
+    for (auto sample = decltype(numSamples){0}; sample < numSamples; sample += 1) {
+        const float t = sampleArray.times[sample];
+
+        auto currentParentMatrices = parentMatrices.Resample(t);
+        auto currentChildMatrices = childMatrices.Resample(t);
 
-    for (auto i = numParentInstances; i > 0; --i) {
-        const auto parentId = i - 1;
-        for (auto j = decltype(numInstances){0}; j < numInstances; ++j) {
-            transforms[j + parentId * numInstances] = transforms[j] * parentTransforms[parentId];
+        sampleArray.values[sample].resize(currentParentMatrices.size() * currentChildMatrices.size());
+        size_t matrix = 0;
+        for (const auto& parentMatrix : currentParentMatrices) {
+            for (const auto& childMatrix : currentChildMatrices) {
+                sampleArray.values[sample][matrix] = childMatrix * parentMatrix;
+                matrix += 1;
+            }
         }
     }
-
-    return transforms;
 }
 
 void HdArnoldInstancer::SetPrimvars(AtNode* node, const SdfPath& prototypeId, size_t instanceCount)
 {
     // TODO(pal): Add support for inheriting primvars from parent instancers.
     VtIntArray instanceIndices;
-    for (const auto& primvar : _primvars) {
-        const auto& desc = primvar.second;
-        if (desc.interpolation != HdInterpolationInstance || !desc.dirtied || primvar.first == _tokens->rotate ||
-            primvar.first == _tokens->translate || primvar.first == _tokens->scale ||
-            primvar.first == _tokens->instanceTransform) {
+    for (auto& primvar : _primvars) {
+        auto& desc = primvar.second;
+        if (!desc.NeedsUpdate()) {
```

Comment:
This is causing primvars to not work when there are multiple prototypes per instancer. We can't rely on the new NeedsUpdate function here.